File: spatialScala/inst/java/spatialScala/src/main/scala/GPs/GP.scala
package spatialScala.GPs

object GP {
  import breeze.linalg.{DenseMatrix, DenseVector, inv}
  import breeze.numerics.{pow, lgamma, exp, log, sqrt}
  import breeze.stats.distributions.{MultivariateGaussian=>MvNorm}
  import spatialScala.mcmc._

  def dist(si: DenseVector[Double], sj: DenseVector[Double]): Double = {
    require(si.length == sj.length)
    val n = si.length

    val diffSq = Vector.tabulate(n){ k => pow(si(k) - sj(k), 2) }
    sqrt( diffSq.sum )
  }

  case class StateMatern(
    beta: DenseVector[Double],
    tau2: Double, // nugget
    sig2: Double, // covariance function scale
    phi: Double,  // range
    nu: Double    // smoothness
  )

  def fitMatern(y: DenseVector[Double],// Observations (Univariate)
          X: DenseMatrix[Double], // Covariates
          s: DenseMatrix[Double], // Locations
          stepSigCov: DenseMatrix[Double] = 
            DenseMatrix.eye[Double](4),
          aTau: Double=2, bTau: Double=1,
          aSig: Double=2, bSig: Double=1,
          aPhi: Double=0, bPhi: Double=1,
          aNu: Double=1.5, bNu: Double=2.5,
          B: Int, burn: Int, printEvery: Int = 0) {

    val (n,k) = (X.rows, X.cols)
    require(y.length == n && s.rows == n)
    
    val D = DenseMatrix.tabulate(n,n){ 
      (i,j) => dist(s(i, ::).t, s(j, ::).t)
    }

    val In = DenseMatrix.eye[Double](n)

    // tau2, sig2, phi, nu
    def trans_param(v: DenseVector[Double]) = {
      DenseVector(
        log(v(0)),
        log(v(1)), 
        Metropolis.logit(v(2), aPhi, bPhi),
        Metropolis.logit(v(3), aNu,  bNu)
      )
    }

    def invTansParam(x: DenseVector[Double]) = {
      DenseVector(
        exp(x(0)),
        exp(x(1)), 
        Metropolis.invLogit(x(2), aPhi, bPhi),
        Metropolis.invLogit(x(3), aNu,  bNu)
      )
    }

    def update(st: StateMatern) = {
      val R = D.map(d => CorrFn.matern(d, st.phi, st.nu))
      val V = st.tau2 * In + st.sig2 * R

      // beta
      val beta = {
        lazy val Vi = inv(V)
        lazy val XtVi = X.t * Vi
        val SigHat = inv(XtVi * X)
        val betaHat = SigHat * XtVi * y
        MvNorm(betaHat, SigHat).sample
      }

      // cov params
      val covParams = {
        def ll(p: DenseVector[Double]) = {
          MvNorm(X*beta, V).logPdf(y)
        }

        def lp(p: DenseVector[Double]) = {
          Metropolis.lpLogInvGamma(p(0), aTau, bTau) + 
          Metropolis.lpLogInvGamma(p(1), aSig, bSig) + 
          Metropolis.lpLogitUnif(p(2)) + 
          Metropolis.lpLogitUnif(p(3)) 
        }

        val curr = DenseVector(
          log(st.tau2),
          log(st.sig2), 
          Metropolis.logit(st.phi, aPhi, bPhi), 
          Metropolis.logit(st.nu, aNu, bNu)
        )

        Metropolis.Multivariate.update(curr, ll, lp, stepSigCov)
      }

      val tau2 = exp(covParams(0))
      val sig2 = exp(covParams(1))
      val phi = Metropolis.invLogit(covParams(2), aPhi, bPhi)
      val nu = Metropolis.invLogit(covParams(3), aNu, bNu)

      StateMatern(beta, tau2, sig2, phi, nu)
    }

    val init = StateMatern(
      beta=DenseVector.zeros[Double](k),
      tau2=1,
      sig2=1,
      phi=(aPhi + bPhi) / 2,
      nu=(aNu + aNu) / 2
    )
    
    val mcmcOut = Gibbs.sample(init, update, B, burn, printEvery)

    lazy val betas = mcmcOut.map(_.beta)
    lazy val betaMat = new DenseMatrix(betas.head.size, betas.size,
      betas.map(_.toArray).toArray.flatten
    ) // k by B

    lazy val covParams = mcmcOut.map(m => 
      Array(m.tau2, m.sig2, m.phi, m.nu)
    )
    lazy val covParamsMat = new DenseMatrix(4, covParams.size,
      covParams.toArray.flatten
    )

    DenseMatrix.horzcat(betaMat, covParamsMat)
  }

}

File: hw/hw4/src/GP_R/gp.R
#geoR::matern(1,2,3)
source("mcmc.R")

matern <- function(d, phi, nu) {
  #' @param
  #' d:     distance
  #' phi:   range
  #' nu:    smoothness
  #' @export
  u <- ifelse(d > 0, d / phi, .Machine$double.eps)
  logR <- -((nu-1) * log(2) + lgamma(nu)) + 
          nu * log(u) + log(besselK(u, nu))
  exp(logR)
}

gp <- function(y, X, s, 
               stepSigCov,
               a_tau=2, b_tau=1, 
               a_sig=2, b_sig=1, 
               a_phi=0, b_phi=2, 
               nu_choice=seq(.5, 2, len=4),
               B=2000, burn=1000, print_every=0) {

  n <- nrow(X)
  k <- ncol(X)
  p <- 3
  Xt <- t(X)
  stopifnot(n == length(y) && n == NROW(s))

  D <- as.matrix(dist(s))
  I_n <- diag(n)

  update <- function(state) {
    out <- state

    # update beta
    R <- matern(d=D, phi=out$phi, nu=out$nu)
    V <- out$tau2 * I_n + out$sig2 * R
    Vi <- solve(V)
    XtVi <- Xt %*% Vi
    Sig_hat <- solve(XtVi %*% X) ### Make sure this is right
    beta_hat <- Sig_hat %*% XtVi %*% y
    out$beta <- mvrnorm(beta_hat, Sig_hat)

    # update tau2, sig2, phi
    ll <- function(trans_param) {
      param <- c(exp(trans_param[1]), 
                 exp(trans_param[2]),
                 inv_logit(trans_param[3], a_phi, b_phi))

      R <- matern(d=D, phi=param[3], nu=out$nu)
      V <- param[1] * I_n + param[2] * R
      
      ldmvnorm(y, X %*% out$beta, V)
    }

    lp <- function(trans_param) {
      lp_log_invgamma(trans_param[1], a_tau, b_tau) +
      lp_log_invgamma(trans_param[2], a_sig, b_sig) +
      lp_logit_unif(trans_param[3])
    }

    trans_curr_cov_param <- c(log(out$tau2), log(out$sig2), 
                              logit(out$phi, a_phi, b_phi))

    mh_trans <- mh_mv(trans_curr_cov_param, ll, lp, stepSigCov)
    out$tau2 <- exp(mh_trans[1])
    out$sig2 <- exp(mh_trans[2])
    out$phi <- inv_logit(mh_trans[3], a_phi, b_phi)

    # Update nu
    ll_nu <- function(nu) {
      R <- matern(d=D, phi=out$phi, nu=nu)
      V <- out$tau2 * I_n + out$sig2 * R
      ldmvnorm(y, X %*% out$beta, V)
    }

    log_prob <- sapply(nu_choice, ll_nu)
    prob <- exp(log_prob - max(log_prob))
    out$nu <- sample(nu_choice, 1, prob=prob)

    out
  }

  init <- list(beta=double(k),
               tau2=b_tau, sig2=b_sig, 
               phi= (a_phi + b_phi) / 2,
               nu= sample(nu_choice, 1))
               #nu = (a_nu + b_nu) / 2)

  gibbs_out <- gibbs(init, update, B, burn, print_every)
  
  out <- matrix(NA, p + k + 1, B) 

  out[1:k, ] <- sapply(gibbs_out, function(x) x$beta)
  out[k+1,] <- sapply(gibbs_out, function(x) x$tau2)
  out[k+2,] <- sapply(gibbs_out, function(x) x$sig2)
  out[k+3,] <- sapply(gibbs_out, function(x) x$phi)
  out[k+4,] <- sapply(gibbs_out, function(x) x$nu)

  rownames(out) <- c(paste0('beta',1:k), 'tau2', 'sig2', 'phi', 'nu')
  t(out)
}


